src/fft.rg

```diff
@@ -48,39 +91,168 @@ fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
 
 local fft = {}
 
-function fft.generate_fft_interface(itype, dtype)
-  assert(regentlib.is_index_type(itype), "requires an index type as the first argument")
-  local dim = itype.dim
+--- Create an FFT interface.
+-- @param itype_input Index type of transform (int1d/int2d/int3d).
+-- @param dtype_in Input data type of transform (float/double/complex32/complex64).
+-- @param dtype_out Output data type of transform (complex32/complex64).
+-- @param batch_flag Flag for batch transform
+function fft.generate_fft_interface(itype_input, dtype_in, dtype_out, batch_flag)
+  assert(regentlib.is_index_type(itype_input), "requires an index type as the first argument")
+  local dim = itype_input.dim
+  local dtype_out_size = terralib.sizeof(dtype_out)
   assert(dim >= 1 and dim <= 3, "currently only 1 <= dim <= 3 is supported")
-  assert(dtype == complex64, "currently only complex64 is supported")
+  assert(dtype_in == float or dtype_in == double or dtype_in == complex32 or dtype_in == complex64, "input type must be float/double/complex32/complex64")
+  assert(dtype_out == complex32 or dtype_out == complex64, "output type must be complex32/complex64")
+
+  local itype = itype_input
+  if batch_flag then
+    dim = dim + 1
+    if itype == int1d then
+      itype = int2d
+    elseif itype == int2d then
+      itype = int3d
+    elseif itype == int3d then
+      itype = int4d
+    end
+  end
+
+   -- Single-precision transforms
+  local float_to_complex32_transform = (dtype_in == float and dtype_out == complex32)
+  local complex32_to_complex32_transform = (dtype_in == complex32 and dtype_out == complex32)
+
+   -- Double-precision transforms
+  local double_to_complex64_transform = (dtype_in == double and dtype_out == complex64)
+  local complex64_to_complex64_transform = (dtype_in == complex64 and dtype_out == complex64)
+
+  -- For precision checks
+  local single_precision = (float_to_complex32_transform or complex32_to_complex32_transform)
+  local double_precision = (double_to_complex64_transform or complex64_to_complex64_transform)
+
+  local fftw_plan_handle_type
+  local fftw_destroy_plan_function
+  if single_precision then
+    regentlib.linklibrary("libfftw3f.so")
+    fftw_plan_handle_type = fftw_c.fftwf_plan
+    fftw_destroy_plan_function = fftw_c.fftwf_destroy_plan
+  elseif double_precision then
+    regentlib.linklibrary("libfftw3.so")
+    fftw_plan_handle_type = fftw_c.fftw_plan
+    fftw_destroy_plan_function = fftw_c.fftw_destroy_plan
```

Comment:
`else assert(false)`

Being pedantic, but might as well be safe here.

Reply:
Edited here: https://github.com/arjunkunna/regent-fft/commit/103303a67e88d24b221b1de468e6d7bf2cc806d5